gpflow/kullback_leiblers.py

```diff
@@ -80,26 +83,49 @@ def gauss_kl(q_mu, q_sqrt, K=None):
         trace = tf.reduce_sum(tf.square(Lq))
     else:
         if diag:
-            M = tf.shape(Lp)[0]
-            Lp_inv = tf.matrix_triangular_solve(
-                Lp, tf.eye(M, dtype=settings.float_type), lower=True)
-            K_inv = tf.matrix_triangular_solve(
-                tf.transpose(Lp), Lp_inv, lower=False)
-            trace = tf.reduce_sum(
-                tf.expand_dims(tf.matrix_diag_part(K_inv), 1) * tf.square(q_sqrt))
+
+            if K.get_shape().ndims == 3:
+                LpT = tf.transpose(Lp,(0,2,1)) # L x M x M
+                Lp_inv = tf.matrix_triangular_solve(Lp,
+                        tf.matrix_diag(tf.ones((L, M), dtype=settings.float_type)), # L x M x M
+                        lower=True) # L x M x M
+                K_inv = tf.transpose(
+                    tf.matrix_diag_part(
+                        tf.matrix_triangular_solve(LpT, Lp_inv, lower=False) # L x M x M
+                    ) # L x M
+                ) # M x L
+            else:
+                LpT = tf.transpose(Lp) # M x M
+                Lp_inv = tf.matrix_triangular_solve(Lp,
+                        tf.eye(M, dtype=settings.float_type), #  M x M
+                        lower=True) # M x M
+                K_inv = tf.matrix_diag_part(
+                    tf.matrix_triangular_solve(LpT, Lp_inv, lower=False)  # M x M
+                )[:,None] # M x 1 (implicit extension to broadcast)
+
+            trace = tf.reduce_sum(K_inv * tf.square(q_sqrt))
+
         else:
-            Lp_tiled = tf.tile(tf.expand_dims(Lp, 0), [num_latent, 1, 1])
-            LpiLq = tf.matrix_triangular_solve(Lp_tiled, Lq, lower=True)
+
+            # If K is L x M x M, no need to tile the cholesky anymore
+            if K.get_shape().ndims == 3:
+                LpiLq = tf.matrix_triangular_solve(Lp, Lq, lower=True)
+            else:
+                Lp_tiled = tf.tile(tf.expand_dims(Lp, 0), [num_latent, 1, 1])
+                LpiLq = tf.matrix_triangular_solve(Lp_tiled, Lq, lower=True)
+
             trace = tf.reduce_sum(tf.square(LpiLq))
 
+
     twoKL = mahalanobis + constant - logdet_qcov + trace
 
     # Log-determinant of the covariance of p(x):
     if not white:
         log_sqdiag_Lp = tf.log(tf.square(tf.matrix_diag_part(Lp)))
         sum_log_sqdiag_Lp = tf.reduce_sum(log_sqdiag_Lp)
-        prior_logdet = tf.cast(num_latent, settings.float_type) * sum_log_sqdiag_Lp
+        # If K is L x M x M, num_latent is no longer implicit, no need to multiply the single kernel logdet
+        scale = tf.cast(L, settings.float_type) if K.get_shape().ndims == 2 else 1.
```

Comment:
`1.` -> `1.0` to be more explicit about the type (my first thought was "huh, you can python like a sentence?"...)